File: src/connectors/git.ts
import { execFile as execFileCb } from "node:child_process";
import { createHash } from "node:crypto";
import { existsSync, readdirSync, readFileSync, statSync } from "node:fs";
import { tmpdir } from "node:os";
import path from "node:path";
import { promisify } from "node:util";
import { isIgnoredPath } from "../fs/paths.js";
import type { ConnectorImpl, ConnectorPollResult } from "./types.js";
import { strConfig, strListConfig } from "./types.js";

const execFile = promisify(execFileCb);

interface GitState {
  commit?: string;
}

/**
 * Git repository connector: shallow-clones once into a temp cache, then
 * fetch/reset on each poll. Emits every matched file; the index pipeline
 * dedupes unchanged documents by content hash.
 */
export const gitConnector: ConnectorImpl = {
  kind: "git",
  async poll(config, state, ctx): Promise<ConnectorPollResult> {
    const url = strConfig(config, "url");
    if (!url) throw new Error("git connector requires config.url");
    const branch = strConfig(config, "branch");
    const extensions = strListConfig(config, "extensions", [".md"]);
    const dir = cacheDir(url);

    await ensureRepo(dir, url, branch);
    const commit = (await git(dir, ["rev-parse", "HEAD"])).trim();
    const previous = (state ?? {}) as GitState;
    if (previous.commit === commit) {
      return { docs: [], state };
    }

    const docs = [];
    for (const filePath of walk(dir, dir)) {
      const ext = path.extname(filePath).toLowerCase();
      if (!extensions.includes(ext)) continue;
      const rel = path.relative(dir, filePath).replace(/\\/g, "/");
      try {
        const stat = statSync(filePath);
        docs.push({
          path: rel,
          content: readFileSync(filePath, "utf8"),
          mtime: Math.floor(stat.mtimeMs),
        });
      } catch (err) {
        ctx.log(`git connector read failed: ${rel}`, err);
      }
    }
    ctx.log(
      `git connector ${url}: ${docs.length} files at ${commit.slice(0, 12)}`,
    );
    return { docs, state: { commit } satisfies GitState };
  },
};

function cacheDir(url: string): string {
  const hash = createHash("sha256").update(url).digest("hex").slice(0, 16);
  return path.join(tmpdir(), `wikillm-git-${hash}`);
}

async function ensureRepo(
  dir: string,
  url: string,
  branch?: string,
): Promise<void> {
  if (existsSync(path.join(dir, ".git"))) {
    if (branch) {
      await git(dir, ["fetch", "--depth", "1", "origin", branch]);
      await git(dir, ["reset", "--hard", "FETCH_HEAD"]);
    } else {
      await git(dir, ["fetch", "--depth", "1", "origin", "HEAD"]);
      await git(dir, ["reset", "--hard", "FETCH_HEAD"]);
    }
    return;
  }
  const args = ["clone", "--depth", "1"];
  if (branch) args.push("--branch", branch);
  args.push(url, dir);
  await git(".", args);
}

async function git(cwd: string, args: string[]): Promise<string> {
  const { stdout } = await execFile("git", args, {
    cwd,
    maxBuffer: 32 * 1024 * 1024,
  });
  return stdout;
}

function walk(root: string, dir: string): string[] {
  const out: string[] = [];
  let entries;
  try {
    entries = readdirSync(dir, { withFileTypes: true });
  } catch {
    return out;
  }
  for (const entry of entries) {
    const full = path.join(dir, entry.name);
    const rel = path.relative(root, full).replace(/\\/g, "/");
    if (isIgnoredPath(rel) || entry.name === ".git") continue;
    if (entry.isDirectory()) out.push(...walk(root, full));
    else if (entry.isFile()) out.push(full);
  }
  return out;
}

File: rust/src/services/connectors/git.rs
//! Git connector: shallow clone into a temp cache, fetch/reset per poll.
//! Emits every matched file; the index pipeline dedupes unchanged content.

use crate::domain::{ConnectorConfig};
use crate::error::{Error, Result};
use serde_json::Value;
use std::path::PathBuf;
use tokio::process::Command;

fn cache_dir(url: &str) -> PathBuf {
    use sha2::Digest;
    let hash = sha2::Sha256::digest(url.as_bytes());
    let hex = hex::encode(hash);
    std::env::temp_dir().join(format!("wikillm-git-{}", &hex[..16]))
}

async fn git(cwd: &std::path::Path, args: &[&str]) -> Result<String> {
    let out = Command::new("git")
        .args(args)
        .current_dir(cwd)
        .output()
        .await
        .map_err(|e| Error::Other(format!("git spawn: {e}")))?;
    if !out.status.success() {
        return Err(Error::Other(format!(
            "git {:?} failed: {}",
            args,
            String::from_utf8_lossy(&out.stderr)
        )));
    }
    Ok(String::from_utf8_lossy(&out.stdout).to_string())
}

fn walk_md(dir: &std::path::Path, root: &std::path::Path, extensions: &[String], out: &mut Vec<(String, String, i64)>) {
    let entries = match std::fs::read_dir(dir) {
        Ok(e) => e,
        Err(_) => return,
    };
    for entry in entries.flatten() {
        let full = entry.path();
        let rel = full.strip_prefix(root).unwrap_or(&full).to_string_lossy().replace('\\', "/");
        if rel == ".git" || rel.starts_with(".git/") {
            continue;
        }
        if full.is_dir() {
            walk_md(&full, root, extensions, out);
        } else if full.is_file() {
            let ext = full.extension().map(|e| format!(".{}", e.to_string_lossy())).unwrap_or_default();
            if extensions.contains(&ext) {
                let mtime = full
                    .metadata()
                    .ok()
                    .and_then(|m| m.modified().ok())
                    .and_then(|t| t.duration_since(std::time::UNIX_EPOCH).ok())
                    .map(|d| d.as_millis() as i64)
                    .unwrap_or(0);
                if let Ok(content) = std::fs::read_to_string(&full) {
                    out.push((rel, content, mtime));
                }
            }
        }
    }
}

/// Returns (docs as (rel_path, title, content, mtime)), new state.
pub async fn poll(config: &Value, state: &Value) -> Result<(Vec<(String, String, String, i64)>, Value)> {
    let url = config.get("url").and_then(|v| v.as_str()).ok_or_else(|| Error::Validation("git connector requires config.url".into()))?.to_string();
    let branch = config.get("branch").and_then(|v| v.as_str());
    let extensions: Vec<String> = config
        .get("extensions")
        .and_then(|v| v.as_array())
        .map(|a| a.iter().filter_map(|v| v.as_str().map(String::from)).collect())
        .unwrap_or_else(|| vec![".md".into()]);
    let dir = cache_dir(&url);

    if dir.join(".git").exists() {
        let fetch_ref = branch.unwrap_or("HEAD");
        git(&dir, &["fetch", "--depth", "1", "origin", fetch_ref]).await?;
        git(&dir, &["reset", "--hard", "FETCH_HEAD"]).await?;
    } else {
        let mut args = vec!["clone", "--depth", "1"];
        if let Some(b) = branch {
            args.extend(["--branch", b]);
        }
        args.push(url.as_str());
        let parent = dir.parent().unwrap_or(std::path::Path::new("."));
        let _ = std::fs::create_dir_all(parent);
        git(parent, &args).await?;
    }
    let commit = git(&dir, &["rev-parse", "HEAD"]).await?.trim().to_string();
    if state.get("commit").and_then(|v| v.as_str()) == Some(commit.as_str()) {
        return Ok((Vec::new(), state.clone()));
    }
    let mut files = Vec::new();
    walk_md(&dir, &dir, &extensions, &mut files);
    files.sort();
    Ok((
        files
            .into_iter()
            .map(|(rel, content, mtime)| (rel.clone(), content, rel, mtime))
            .map(|(a, b, c, d)| (a, b, c, d))
            .collect(),
        serde_json::json!({ "commit": commit }),
    ))
}

// ConnectorConfig re-export guard for signature parity with TS naming.
#[allow(dead_code)]
fn _cfg_shape(_: ConnectorConfig) {}
